client/checkout/api/index.js

```diff
@@ -156,10 +156,7 @@ export default class WCPayAPI {
 		const isSetupIntent = 'si' === partials[ 1 ];
 		let orderId = partials[ 2 ];
 		const clientSecret = partials[ 3 ];
-		// Update the current order status nonce with the new one to ensure that the update
-		// order status call works when a guest user creates an account during checkout.
-		// eslint-disable-next-line camelcase
-		setConfig( 'updateOrderStatusNonce', partials[ 4 ] );
```

Comment:
I think there were a reason for doing it so complex. Do you know why it was implemented like that before? I'm afraid that nonce value might be unavailable in some cases unless stored in wcSettings. 

I'm also curious if deprecation of `setSettings` method was done without recommended replacement?

Reply:
My best guess is that it was designed to always have [nonces](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payment-gateway-wcpay.php#L467-L468) available in global variable and when nonce for updating the order status needs to be updated, in this case, only when 3ds credit card is used and when user is a guest creating an account (because nonce depends on user id and guest and logged in user id have different user id), it's cleaner to just update the global variable with the new nonce and read the nonce value from global variable when updating the order status.

We can stick with that design but need to be a bit hacky by modifying the value of the global variable `wcSettings.woocommerce_payments_data.updateOrderStatusNonce` directly. It's mentioned [here](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/1838) that though wcSettings should be immutable but immutability can't be enforced.

Reply:
I read the point about immutability and get it. Then there should be a way to add the nonce value to the wcSettings on the site via filter or something else, no? If so, I'd take this approach and keep reading nonce from the settings rather than relying on js to set nonce.

> My best guess is that it was designed to always have nonces available in global variable and when nonce for updating the order status needs to be updated, in this case, only when 3ds credit card is used and when user is a guest creating an account (because nonce depends on user id and guest and logged in user id have different user id), it's cleaner to just update the global variable with the new nonce and read the nonce value from global variable when updating the order status.

It's a good guess, but I'd ask authors. @RadoslavGeorgiev , @DanReyLop and @dwainm could please add more info about nonces in checkout blocks? Why did we set it via JS rather than PHP?

Reply:
@shendy-a8c is on the right track.

1. `updateOrderStatusNonce` is defined in PHP [here](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payment-gateway-wcpay.php#L467).
2. It __will__ be updated [here](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/api/index.js#L161) in case the server sends a new one after a redirect URL has been generated (remember, they carry order IDs, client secret, etc.). At the point of writing the code, I have seemingly thought that this step is optional, but it isn't: it's always happening.
3. And it's finally used [here](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/api/index.js#L204).

Those are all the 3 places where `updateOrderStatusNonce` appears.

Looking at [the original PR](https://github.com/Automattic/woocommerce-payments/pull/723/files#diff-bc05e644f9133795b10f0696ea2862c0abeecfeed6f3ff2eaf48a691c701e14bR149), the flow is already the same as described in the list above. At this point I'm inclined to think that this is just a mistake from my side, and that including the nonce in `get_payment_fields_js_config` is not needed at all, since it's received along with the intent data.

---

>  This can only be tested with shortcode checkout because I don't see checkout block lets you create an account.

@shendy-a8c please take a look at this comment: https://github.com/Automattic/woocommerce-payments/pull/723/files#r494208543
You can test this with the checkout block as well, although being in the `api/index.js` file, the change will probably not break the block.

Reply:
Thank you @RadoslavGeorgiev for clarifying. My next step would be to remove `updateOrderStatusNonce` from `get_payment_fields_js_config` and test guest checkout creating an account in checkout blocks and shortcode checkout.

I saw that [PR](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2851) already merged but I didn't see the option to create an account in checkout block. I now realize that the toggle to turn that on is on the block's settings, not in Woo's settings.